engine/src/main/java/org/entando/entando/aps/system/services/oauth2/OAuth2TokenDAO.java

```diff
@@ -218,5 +245,45 @@ public void deleteExpiredToken() {
             closeDaoResources(null, stat, conn);
         }
     }
-
+    
+    @Override
+    public OAuth2RefreshToken readRefreshToken(String tokenValue) {
+        FieldSearchFilter filter = new FieldSearchFilter("refreshtoken", tokenValue, true);
+        FieldSearchFilter[] filters = {filter};
+        List<String> accessTokens = super.searchId(filters);
+        if (null != accessTokens && accessTokens.size() > 0) {
+            return new DefaultOAuth2RefreshToken(tokenValue);
+        }
+        return null;
+    }
+    
+    @Override
+    public OAuth2Authentication readAuthenticationForRefreshToken(OAuth2RefreshToken refreshToken) {
+        OAuth2Authentication authentication = null;
+        Connection conn = null;
+        PreparedStatement stat = null;
+        ResultSet res = null;
+        try {
+            conn = this.getConnection();
+            stat = conn.prepareStatement(SELECT_TOKEN_BY_REFRESH);
+            stat.setString(1, refreshToken.getValue());
+            res = stat.executeQuery();
+            if (res.next()) {
+                String username = res.getString("localuser");
+                UsernamePasswordAuthenticationToken auth = new UsernamePasswordAuthenticationToken(username, "");
+                String clientid = res.getString("clientid");
+                Map<String, String> requestParameters = new HashMap<>();
+                requestParameters.put(OAuth2Utils.GRANT_TYPE, res.getString("granttype"));
+                OAuth2Request oAuth2Request = new OAuth2Request(requestParameters, clientid, null, true, null, null, null, null, null);
+                authentication = new OAuth2Authentication(oAuth2Request, auth);
+            }
+        } catch (Exception t) {
+            logger.error("Error while reading tokens", t);
+            throw new RuntimeException("Error while reading tokens", t);
```

Comment:
Can we throw something more specific here? Not sure if it is totally required or not but a generic "throw RuntimeException" isn't great if avoidable

Reply:
Correct, we can throw something specific for the context (like an SQLException or something like that) but we need to decide if change it only here or apply a general strategy in every dao in the platform (now, in ours DAOs, we are using RuntimeException without any "throws" in method declaration)

Reply:
We can update these as we encounter them or make changes. In general, we should prefer specific exceptions

Reply:
Reviewed with Eugenio and we'll leave this for now. Throwing a more specific exception trickles all the way back to an implementation of a spring interface. We would need to rework to much of the infrastructure to update it.